File: crates/qualification-driver/src/scenario/error_envelope.rs
use std::sync::Arc;

use super::{ActionOutcome, Runner};
use guardian_client::{FalconKeyStore, GuardianClient};
use miden_protocol::account::AccountId;

/// Syntactically valid and deliberately never registered. The account lookup
/// runs before timestamp validation and before signature verification, so a
/// throwaway credential still reaches the envelope this asserts.
const UNREGISTERED_ACCOUNT_ID: &str = "0xaabbccddeeff00011b27a8df4ddbe0";
const EXPECTED_CODE: &str = "account_not_found";

pub async fn assert_grpc_envelope(runner: &Runner) -> ActionOutcome {
    let account_id = match AccountId::from_hex(UNREGISTERED_ACCOUNT_ID) {
        Ok(id) => id,
        Err(error) => {
            return ActionOutcome::failed_setup(format!(
                "the fixture account id is malformed: {error}"
            ));
        }
    };

    // Credentials are extracted before the account is looked up, so an
    // unauthenticated request is rejected at the boundary and never reaches the
    // envelope this asserts. A throwaway signer gets past that.
    let signer = Arc::new(FalconKeyStore::generate());
    let mut client = match GuardianClient::connect(runner.endpoints.grpc.clone()).await {
        Ok(client) => client.with_signer(signer),
        Err(error) => {
            return ActionOutcome::failed_setup(format!(
                "cannot reach {}: {error}",
                runner.endpoints.grpc
            ));
        }
    };

    match client.get_state(&account_id).await {
        Ok(_) => ActionOutcome::failed_product(
            "an unregistered account returned state instead of an error".to_string(),
        ),
        Err(error) => match error.guardian_code() {
            Some(code) if code == EXPECTED_CODE => ActionOutcome::Passed,
            Some(code) => ActionOutcome::failed_product(format!(
                "expected code `{EXPECTED_CODE}` over gRPC, got `{code}`"
            )),
            None => ActionOutcome::failed_product(format!(
                "gRPC error carried no structured envelope: {error}"
            )),
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// A malformed fixture would make the assertion fail while parsing instead
    /// of while reading the envelope, which looks like a product defect.
    #[test]
    fn the_unregistered_account_id_is_structurally_valid() {
        AccountId::from_hex(UNREGISTERED_ACCOUNT_ID)
            .expect("the unregistered fixture account id must parse");
    }
}

File: packages/miden-multisig-client/tests/qualification/actions/errorEnvelope.ts
import type { ActionOutcome, ActionContext } from '../runner.js';

/**
 * Syntactically valid and deliberately never registered. The account lookup
 * runs before timestamp validation and before signature verification, so a
 * throwaway credential still reaches the envelope this asserts.
 */
const UNREGISTERED_ACCOUNT_ID = '0xaabbccddeeff00011b27a8df4ddbe0';
const EXPECTED_CODE = 'account_not_found';

interface ErrorBody {
  code: string;
  message: string;
  meta: { retryable: boolean };
}

export async function assertHttpEnvelope(context: ActionContext): Promise<ActionOutcome> {
  const url = `${context.httpEndpoint}/state?account_id=${UNREGISTERED_ACCOUNT_ID}`;
  let response: Response;
  try {
    response = await fetch(url, {
      headers: {
        'x-pubkey': '00'.repeat(32),
        'x-signature': '00'.repeat(64),
        'x-timestamp': String(Math.floor(Date.now() / 1000)),
      },
    });
  } catch (error) {
    return { kind: 'failed', classification: 'setup', reason: `GET ${url} failed: ${String(error)}` };
  }

  let body: ErrorBody;
  try {
    body = (await response.json()) as ErrorBody;
  } catch (error) {
    return {
      kind: 'failed',
      classification: 'product',
      reason: `error response carried no structured envelope: ${String(error)}`,
    };
  }

  if (response.status !== 404) {
    return {
      kind: 'failed',
      classification: 'product',
      reason: `expected 404 for an unregistered account, got ${response.status} with code ${body.code}`,
    };
  }
  if (body.code !== EXPECTED_CODE) {
    return { kind: 'failed', classification: 'product', reason: `expected code ${EXPECTED_CODE}, got ${body.code}` };
  }
  if (!body.message?.trim()) {
    return { kind: 'failed', classification: 'product', reason: 'envelope carried an empty message' };
  }
  if (body.meta?.retryable) {
    return { kind: 'failed', classification: 'product', reason: 'an unregistered account was reported as retryable' };
  }
  return { kind: 'passed' };
}
